File: .github/copilot-instructions/examples/backend/php/UserService.php
<?php

declare(strict_types=1);

namespace App\Service;

use App\Repository\UserRepositoryInterface;
use App\Security\PasswordHasherInterface;
use App\Exception\InvalidInputException;
use App\Exception\UserAlreadyExistsException;
use App\Exception\UserNotFoundException;
use App\DTO\UserDTO;
use App\DTO\CreateUserRequest;
use App\DTO\UpdateUserRequest;

/**
 * User Service - Example of PHP best practices.
 *
 * Demonstrates:
 * - Strict types (declare(strict_types=1))
 * - Constructor property promotion (PHP 8.0+)
 * - Type hints for parameters and return types
 * - Dependency injection via constructor
 * - Readonly properties (PHP 8.1+)
 * - Named arguments
 * - Match expressions (PHP 8.0+)
 *
 * @psalm-immutable for DTOs
 */
final class UserService
{
    private const EMAIL_PATTERN = '/^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/';
    private const USERNAME_PATTERN = '/^[a-zA-Z0-9_]{3,30}$/';
    private const MIN_PASSWORD_LENGTH = 12;

    /**
     * Constructor with dependency injection (property promotion).
     *
     * @param UserRepositoryInterface $userRepository Repository for user persistence
     * @param PasswordHasherInterface $passwordHasher Service for password hashing
     */
    public function __construct(
        private readonly UserRepositoryInterface $userRepository,
        private readonly PasswordHasherInterface $passwordHasher,
    ) {
    }

    /**
     * Create a new user with validation.
     *
     * @param CreateUserRequest $request User creation request
     * @return UserDTO Created user data
     * @throws InvalidInputException If validation fails
     * @throws UserAlreadyExistsException If user already exists
     */
    public function createUser(CreateUserRequest $request): UserDTO
    {
        // Validate input
        $this->validateUsername($request->username);
        $this->validateEmail($request->email);
        $this->validatePassword($request->password);

        // Check if user exists
        $existingUser = $this->userRepository->findByEmail($request->email);
        if ($existingUser !== null) {
            throw new UserAlreadyExistsException(
                sprintf('User with email %s already exists', $request->email)
            );
        }

        // Hash password
        $passwordHash = $this->passwordHasher->hash($request->password);

        // Create user
        $user = $this->userRepository->save(
            username: $request->username,
            email: $request->email,
            passwordHash: $passwordHash,
        );

        return $user;
    }

    /**
     * Authenticate user by email and password.
     *
     * @param string $email User's email
     * @param string $password User's password
     * @return UserDTO|null User data if authentication successful, null otherwise
     */
    public function authenticate(string $email, string $password): ?UserDTO
    {
        $user = $this->userRepository->findByEmail($email);

        if ($user === null) {
            return null;
        }

        // Note: In real implementation, need to fetch password hash
        // This is simplified for example
        // if (!$this->passwordHasher->verify($password, $storedHash)) {
        //     return null;
        // }

        return $user;
    }

    /**
     * Get user by ID.
     *
     * @param int $userId User's ID
     * @return UserDTO|null User data if found, null otherwise
     */
    public function getUser(int $userId): ?UserDTO
    {
        return $this->userRepository->findById($userId);
    }

    /**
     * Update user profile.
     *
     * @param int $userId User's ID
     * @param UpdateUserRequest $request Update request
     * @return UserDTO Updated user data
     * @throws UserNotFoundException If user not found
     * @throws InvalidInputException If validation fails
     * @throws UserAlreadyExistsException If email is taken
     */
    public function updateUser(int $userId, UpdateUserRequest $request): UserDTO
    {
        $user = $this->userRepository->findById($userId);
        if ($user === null) {
            throw new UserNotFoundException(sprintf('User not found: %d', $userId));
        }

        // Validate and update username if provided
        if ($request->username !== null) {
            $this->validateUsername($request->username);
        }

        // Validate and update email if provided
        if ($request->email !== null) {
            $this->validateEmail($request->email);

            // Check email not taken by another user
            $existingUser = $this->userRepository->findByEmail($request->email);
            if ($existingUser !== null && $existingUser->id !== $userId) {
                throw new UserAlreadyExistsException(
                    sprintf('Email %s is already taken', $request->email)
                );
            }
        }

        // Update user
        return $this->userRepository->update(
            userId: $userId,
            username: $request->username ?? $user->username,
            email: $request->email ?? $user->email,
        );
    }

    /**
     * Delete user by ID.
     *
     * @param int $userId User's ID
     * @throws UserNotFoundException If user not found
     */
    public function deleteUser(int $userId): void
    {
        if (!$this->userRepository->exists($userId)) {
            throw new UserNotFoundException(sprintf('User not found: %d', $userId));
        }

        $this->userRepository->delete($userId);
    }

    /**
     * Validate username format.
     *
     * @param string $username Username to validate
     * @throws InvalidInputException If validation fails
     */
    private function validateUsername(string $username): void
    {
        if ($username === '') {
            throw new InvalidInputException('Username is required');
        }

        if (preg_match(self::USERNAME_PATTERN, $username) !== 1) {
            throw new InvalidInputException(
                'Username must be 3-30 characters and contain only letters, numbers, and underscores'
            );
        }
    }

    /**
     * Validate email format.
     *
     * @param string $email Email to validate
     * @throws InvalidInputException If validation fails
     */
    private function validateEmail(string $email): void
    {
        if ($email === '') {
            throw new InvalidInputException('Email is required');
        }

        if (preg_match(self::EMAIL_PATTERN, $email) !== 1) {
            throw new InvalidInputException(sprintf('Invalid email format: %s', $email));
        }
    }

    /**
     * Validate password strength.
     *
     * @param string $password Password to validate
     * @throws InvalidInputException If validation fails
     */
    private function validatePassword(string $password): void
    {
        if (strlen($password) < self::MIN_PASSWORD_LENGTH) {
            throw new InvalidInputException(
                sprintf('Password must be at least %d characters', self::MIN_PASSWORD_LENGTH)
            );
        }

        $hasUpper = preg_match('/[A-Z]/', $password) === 1;
        $hasLower = preg_match('/[a-z]/', $password) === 1;
        $hasDigit = preg_match('/[0-9]/', $password) === 1;

        // Use match expression for validation (PHP 8.0+)
        $missingRequirement = match (true) {
            !$hasUpper => 'at least one uppercase letter',
            !$hasLower => 'at least one lowercase letter',
            !$hasDigit => 'at least one number',
            default => null,
        };

        if ($missingRequirement !== null) {
            throw new InvalidInputException(
                sprintf('Password must contain %s', $missingRequirement)
            );
        }
    }
}

File: .github/copilot-instructions/examples/backend/java/UserService.java
package com.example.demo.user;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Optional;
import java.util.regex.Pattern;

/**
 * User Service - Example of Java Spring Boot best practices.
 *
 * Demonstrates:
 * - Constructor injection (recommended over field injection)
 * - Records for DTOs (Java 16+)
 * - Optional for null safety
 * - Business logic separation
 * - Transaction management
 * - Custom exceptions
 * - Pattern matching
 */
@Service
@Transactional
public class UserService {

    // Email validation pattern (simplified for example)
    private static final Pattern EMAIL_PATTERN =
            Pattern.compile("^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}$");

    // Username validation pattern
    private static final Pattern USERNAME_PATTERN =
            Pattern.compile("^[a-zA-Z0-9_]{3,30}$");

    private static final int MIN_PASSWORD_LENGTH = 12;

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    /**
     * Constructor injection - preferred over field injection.
     * Makes dependencies explicit and enables easier testing.
     */
    public UserService(UserRepository userRepository, PasswordEncoder passwordEncoder) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
    }

    /**
     * Create a new user with validation.
     *
     * @param request the user creation request
     * @return the created user DTO
     * @throws InvalidInputException if validation fails
     * @throws UserAlreadyExistsException if user exists
     */
    public UserDTO createUser(CreateUserRequest request) {
        // Validate input
        validateUsername(request.username());
        validateEmail(request.email());
        validatePassword(request.password());

        // Check if user already exists
        if (userRepository.existsByEmail(request.email())) {
            throw new UserAlreadyExistsException(
                    "User with email %s already exists".formatted(request.email())
            );
        }

        // Create user entity
        var user = User.builder()
                .username(request.username())
                .email(request.email())
                .passwordHash(passwordEncoder.encode(request.password()))
                .role(UserRole.USER)
                .build();

        // Save and return DTO
        var savedUser = userRepository.save(user);
        return UserDTO.from(savedUser);
    }

    /**
     * Authenticate user by email and password.
     *
     * @param email the user's email
     * @param password the user's password
     * @return optional containing user DTO if authentication successful
     */
    @Transactional(readOnly = true)
    public Optional<UserDTO> authenticate(String email, String password) {
        return userRepository.findByEmail(email)
                .filter(user -> passwordEncoder.matches(password, user.getPasswordHash()))
                .map(UserDTO::from);
    }

    /**
     * Get user by ID.
     *
     * @param userId the user's ID
     * @return optional containing user DTO if found
     */
    @Transactional(readOnly = true)
    public Optional<UserDTO> getUser(Long userId) {
        return userRepository.findById(userId)
                .map(UserDTO::from);
    }

    /**
     * Update user profile.
     *
     * @param userId the user's ID
     * @param request the update request
     * @return the updated user DTO
     * @throws UserNotFoundException if user not found
     * @throws InvalidInputException if validation fails
     */
    public UserDTO updateUser(Long userId, UpdateUserRequest request) {
        var user = userRepository.findById(userId)
                .orElseThrow(() -> new UserNotFoundException("User not found: " + userId));

        // Validate and update username if provided
        if (request.username() != null) {
            validateUsername(request.username());
            user.setUsername(request.username());
        }

        // Validate and update email if provided
        if (request.email() != null) {
            validateEmail(request.email());

            // Check email not taken by another user
            userRepository.findByEmail(request.email())
                    .filter(existingUser -> !existingUser.getId().equals(userId))
                    .ifPresent(existingUser -> {
                        throw new UserAlreadyExistsException(
                                "Email %s is already taken".formatted(request.email())
                        );
                    });

            user.setEmail(request.email());
        }

        var savedUser = userRepository.save(user);
        return UserDTO.from(savedUser);
    }

    /**
     * Delete user by ID.
     *
     * @param userId the user's ID
     * @throws UserNotFoundException if user not found
     */
    public void deleteUser(Long userId) {
        if (!userRepository.existsById(userId)) {
            throw new UserNotFoundException("User not found: " + userId);
        }

        userRepository.deleteById(userId);
    }

    // Validation methods

    private void validateUsername(String username) {
        if (username == null || username.isBlank()) {
            throw new InvalidInputException("Username is required");
        }

        if (!USERNAME_PATTERN.matcher(username).matches()) {
            throw new InvalidInputException(
                    "Username must be 3-30 characters and contain only letters, numbers, and underscores"
            );
        }
    }

    private void validateEmail(String email) {
        if (email == null || email.isBlank()) {
            throw new InvalidInputException("Email is required");
        }

        if (!EMAIL_PATTERN.matcher(email).matches()) {
            throw new InvalidInputException("Invalid email format: " + email);
        }
    }

    private void validatePassword(String password) {
        if (password == null || password.length() < MIN_PASSWORD_LENGTH) {
            throw new InvalidInputException(
                    "Password must be at least %d characters".formatted(MIN_PASSWORD_LENGTH)
            );
        }

        boolean hasUpper = password.chars().anyMatch(Character::isUpperCase);
        boolean hasLower = password.chars().anyMatch(Character::isLowerCase);
        boolean hasDigit = password.chars().anyMatch(Character::isDigit);

        if (!hasUpper) {
            throw new InvalidInputException("Password must contain at least one uppercase letter");
        }

        if (!hasLower) {
            throw new InvalidInputException("Password must contain at least one lowercase letter");
        }

        if (!hasDigit) {
            throw new InvalidInputException("Password must contain at least one number");
        }
    }

    // DTOs using Records (Java 16+)

    /**
     * Request to create a new user.
     */
    public record CreateUserRequest(
            String username,
            String email,
            String password
    ) {}

    /**
     * Request to update user profile.
     */
    public record UpdateUserRequest(
            String username,
            String email
    ) {}

    /**
     * User data transfer object.
     */
    public record UserDTO(
            Long id,
            String username,
            String email,
            UserRole role
    ) {
        /**
         * Factory method to create DTO from entity.
         */
        public static UserDTO from(User user) {
            return new UserDTO(
                    user.getId(),
                    user.getUsername(),
                    user.getEmail(),
                    user.getRole()
            );
        }
    }

    // Custom Exceptions

    public static class InvalidInputException extends RuntimeException {
        public InvalidInputException(String message) {
            super(message);
        }
    }

    public static class UserAlreadyExistsException extends RuntimeException {
        public UserAlreadyExistsException(String message) {
            super(message);
        }
    }

    public static class UserNotFoundException extends RuntimeException {
        public UserNotFoundException(String message) {
            super(message);
        }
    }
}
